actioncable/lib/action_cable/connection/web_socket_rack.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module ActionCable
+  module Connection
+    class WebSocket
+      # Wrap the `rack.upgrade?` API as an alternative to using `websocket/driver` and `nio4r`
+      #
+      # The `rack.upgrade?` approach detailed here: https://github.com/rack/rack/pull/1272
+      class WebSocketRack # :nodoc:
+        UPGRADE_EXISTS = "rack.upgrade?"
+        UPGRADE = "rack.upgrade"
+        PROTOCOL_NAME_IN = "HTTP_SEC_WEBSOCKET_PROTOCOL"
+        PROTOCOL_NAME_OUT = "Sec-Websocket-Protocol"
+        CLOSE_REASON = ""
+
+        attr_reader :protocol
+
+        def initialize(env, event_target, event_loop, protocols)
+          env[UPGRADE] = self
+          @event_target = event_target
+          @protocol = nil
+          @websocket = nil
+          request_protocols = env[PROTOCOL_NAME_IN]
+          return if request_protocols.nil?
+          request_protocols = request_protocols.split(/,\s?/) if request_protocols.is_a?(String)
+          request_protocols.each do |request_protocol|
+            break(@protocol = request_protocol) if protocols.include?(request_protocol)
+          end
+        end
+
+        def alive?
+          websocket&.open?
+        end
+
+        def transmit(data)
+          return false unless websocket
+          case data
+          when Numeric then websocket.write(data.to_s)
+          when String  then websocket.write(data)
+          when Array   then websocket.write(data.pack("C*"))
+          else false
+          end
+        end
+
+        def close
+          websocket&.close
+        end
+
+        def rack_response
+          return [101, { PROTOCOL_NAME_OUT => protocol }, []] if protocol
+          [101, {}, []]
```

Comment:
I'll can this, since you asked, but I don't really understand... early returns are good for both performance and readability.

Maybe it's the C programmer in me, but early returns minimize code jumps and the compiler will often add early returns if possible (to avoid multiple jumps).

Reply:
Early returns require reading code in imperative style, scanning for exit points. Ruby generally encourages reading code as expressions, so they can be absorbed as clauses rather than examined line by line.

Guard clauses are a notable exception, because their presence at the beginning of a method communicates something meaningful about the method's intent & purpose, rather than its execution alone: that a precondition must be met.